File: electron/main/rules/create.ts
import fs from 'node:fs/promises'
import path from 'node:path'
import { getAdapter } from './adapters'
import { getTemplateContent, templateFileName, type TemplateId } from './templates'
import {
  makeAssetId,
  hashContent,
  normalizeSlug,
  titleFromContent,
  type AssetKind,
  type RuleAsset,
  type ToolId,
} from './types'

export interface CreateRuleOptions {
  toolId: ToolId
  template: TemplateId
  slug?: string
  /** 覆盖推断的 kind；默认由模板决定 */
  kind?: AssetKind
}

function kindForTemplate(template: TemplateId): AssetKind {
  if (template === 'skill') return 'skill'
  if (template === 'mdc') return 'rule'
  return 'rule'
}

export async function createRuleAsset(options: CreateRuleOptions): Promise<RuleAsset> {
  const adapter = getAdapter(options.toolId)
  if (!adapter) throw new Error('未知工具')

  const roots = adapter.getRoots().map((r) => path.resolve(r))
  let root = roots[0]
  for (const r of roots) {
    try {
      await fs.access(r)
      root = r
      break
    } catch {
      // try next
    }
  }
  if (!root) throw new Error('未找到工具配置目录')

  try {
    await fs.mkdir(root, { recursive: true })
  } catch {
    // ignore
  }

  const slug = options.slug?.trim() || 'untitled'
  const { relDir, fileName } = templateFileName(options.template, slug)
  const dir = relDir ? path.join(root, relDir) : root
  await fs.mkdir(dir, { recursive: true })

  const absPath = path.join(dir, fileName)
  let exists = false
  try {
    await fs.access(absPath)
    exists = true
  } catch {
    exists = false
  }
  if (exists) {
    throw new Error(`文件已存在：${path.relative(root, absPath)}`)
  }

  const content = getTemplateContent(options.template, slug)
  await fs.writeFile(absPath, content, 'utf8')
  const st = await fs.stat(absPath)
  const relPath = path.relative(root, absPath)
  const kind = options.kind || kindForTemplate(options.template)

  return {
    id: makeAssetId(options.toolId, absPath),
    toolId: options.toolId,
    kind,
    title: titleFromContent(fileName, relPath, content),
    slug: normalizeSlug(fileName, relPath),
    absPath,
    relPath,
    rootPath: root,
    size: st.size,
    mtime: st.mtimeMs,
    contentHash: hashContent(content),
  }
}

File: src-tauri/src/rules/create.rs
use std::fs;
use std::path::PathBuf;

use super::adapters::{get_adapter, resolve_root_path};
use super::templates::{get_template_content, template_file_name};
use super::types::{
  hash_content, make_asset_id, mtime_ms, normalize_slug, title_from_content, AssetKind, RuleAsset,
};

fn kind_for_template(template: &str) -> AssetKind {
  if template == "skill" {
    "skill".into()
  } else {
    "rule".into()
  }
}

pub fn create_rule_asset(
  tool_id: &str,
  template: &str,
  slug: Option<&str>,
  kind: Option<&str>,
) -> Result<RuleAsset, String> {
  let adapter = get_adapter(tool_id).ok_or("未知工具")?;
  let roots: Vec<PathBuf> = (adapter.roots)()
    .into_iter()
    .map(|r| r.canonicalize().unwrap_or(r))
    .collect();

  let mut root = roots.first().cloned().ok_or("未找到工具配置目录")?;
  for r in &roots {
    if r.exists() {
      root = r.clone();
      break;
    }
  }

  let _ = fs::create_dir_all(&root);
  let slug_str = slug.unwrap_or("untitled").trim();
  let (rel_dir, file_name) = template_file_name(template, slug_str);
  let dir = if rel_dir.is_empty() {
    root.clone()
  } else {
    root.join(&rel_dir)
  };
  fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
  let abs_path = dir.join(&file_name);
  if abs_path.exists() {
    let rel = abs_path.strip_prefix(&root).unwrap_or(&abs_path);
    return Err(format!("文件已存在：{}", rel.to_string_lossy()));
  }

  let content = get_template_content(template, slug_str);
  fs::write(&abs_path, &content).map_err(|e| e.to_string())?;
  let meta = fs::metadata(&abs_path).map_err(|e| e.to_string())?;
  let rel_path = abs_path
    .strip_prefix(&root)
    .unwrap_or(&abs_path)
    .to_string_lossy()
    .to_string();
  let kind = kind
    .map(String::from)
    .unwrap_or_else(|| kind_for_template(template));
  let abs_str = abs_path.to_string_lossy().to_string();

  Ok(RuleAsset {
    id: make_asset_id(tool_id, &abs_str),
    tool_id: tool_id.to_string(),
    kind,
    title: title_from_content(&file_name, &rel_path, &content),
    slug: normalize_slug(&file_name, &rel_path),
    abs_path: abs_str,
    rel_path,
    root_path: resolve_root_path(&root),
    size: meta.len(),
    mtime: mtime_ms(&meta),
    content_hash: Some(hash_content(content.as_bytes())),
  })
}
